packages/async-flow/src/endowments.js

```diff
@@ -0,0 +1,261 @@
+ * @import {Callable} from '@agoric/internal'
+ * @import {PreparationOptions} from '../src/types.js'
+ */
+
+/**
+ * @typedef {'promise' | 'storable' | 'far' | 'function' | 'array' | 'record' | 'state'} EndowmentKind
+ */
+
+const {
+  getOwnPropertyDescriptor,
+  getOwnPropertyDescriptors,
+  create,
+  fromEntries,
+  entries,
+  prototype: objectPrototype,
+} = Object;
+const { ownKeys } = Reflect;
+
+const FunctionWrapperI = M.interface('FunctionWrapper', {
+  apply: M.call(M.array()).returns(M.any()),
+});
+
+const StateAccessorI = M.interface('StateAccessor', {
+  get: M.call(PropertyKeyShape).returns(M.any()),
+  set: M.call(PropertyKeyShape, M.any()).returns(),
+});
+
+const UnwrapperI = M.interface('Unwrapper', {
+  unwrap: M.call(M.remotable('guestWrapped')).returns(M.raw()),
+});
+
+export const forwardingMethods = rem => {
+  const keys = rem[GET_METHOD_NAMES]();
+  const makeMethodEntry = key =>
+    entries({
+      [key](...args) {
+        return rem[key](...args);
+      },
+    })[0];
+  return fromEntries(keys.map(makeMethodEntry));
+};
+
+/**
+ * @param {Zone} outerZone
+ * @param {PreparationOptions} [outerOptions]
+ */
+export const prepareEndowmentTools = (outerZone, outerOptions = {}) => {
+  const { vowTools = prepareVowTools(outerZone) } = outerOptions;
+  const { makeVowKit } = vowTools;
+
+  const functionUnwrapper = outerZone.exo('FunctionUnwrapper', UnwrapperI, {
+    unwrap(guestWrapped) {
+      const unwrapped = (...args) => guestWrapped.apply(args);
+      return harden(unwrapped);
+    },
+  });
+
+  const makeStateUnwrapper = outerZone.exoClass(
+    'StateUnwrapper',
+    UnwrapperI,
+    keys => ({ keys }),
+    {
+      unwrap(guestWrapped) {
+        const { state } = this;
+        const { keys } = state;
+        return harden(
+          create(
+            objectPrototype,
+            fromEntries(
+              keys.flatMap(key =>
+                entries(
+                  getOwnPropertyDescriptors({
+                    get [key]() {
+                      return guestWrapped.get(key);
+                    },
+                    set [key](newValue) {
+                      guestWrapped.set(key, newValue);
+                    },
+                  }),
+                ),
+              ),
+            ),
+          ),
+        );
+      },
+    },
+  );
+
+  /**
+   * Endowment taxonomy. Expected to grow over time.
+   * Defined within `prepareEndowmentTools` because isStorable depends on zone.
+   *
+   * @param {unknown} e
+   * @returns {EndowmentKind}
+   */
+  const endowmentKindOf = e => {
+    harden(e);
+    if (isPromise(e)) {
+      return 'promise';
+    } else if (outerZone.isStorable(e)) {
+      return 'storable';
+    } else if (isPassable(e) && isRemotable(e)) {
+      return 'far';
+    } else if (typeof e === 'function') {
+      return 'function';
+    } else if (typeof e === 'object') {
+      if (e === null) {
+        throw Fail`internal: null is always storable`;
```

Comment:
Does this mean it shouldn't be passed as an endowment?  Because you might have an endowment that is e.g., optional, and therefore could in production be null.

Reply:
Oh. I think you mean it should be already addressed in the 'storable' case above

Reply:
> Oh. I think you mean it should be already addressed in the 'storable' case above

Correct. That's why the error message begins "internal:".

Resolving.